sktime/dists_kernels/dtw.py

```diff
@@ -0,0 +1,171 @@
+        difference.
+
+    References
+    ----------
+    .. [1] H. Sakoe, S. Chiba, "Dynamic programming algorithm optimization for
+           spoken word recognition," IEEE Transactions on Acoustics, Speech and
+           Signal Processing, vol. 26(1), pp. 43--49, 1978.
+    .. [2] Keogh, Eamonn & Pazzani, Michael. (2002). Derivative Dynamic Time Warping.
+        First SIAM International Conference on Data Mining.
+        1. 10.1137/1.9781611972719.1.
+    .. [3] Young-Seon Jeong, Myong K. Jeong, Olufemi A. Omitaomu, Weighted dynamic time
+    warping for time series classification, Pattern Recognition, Volume 44, Issue 9,
+    2011, Pages 2231-2240, ISSN 0031-3203, https://doi.org/10.1016/j.patcog.2010.09.022.
+    """
+
+    _tags = {
+        "symmetric": True,  # all the distances are symmetric
+        "X_inner_mtype": "numpy3D",
+    }
+
+    def __init__(
+        self,
+        weighted: bool = False,
+        derivative: bool = False,
+        window: Union[int, None] = None,
+        itakura_max_slope: Union[float, None] = None,
+        bounding_matrix: np.ndarray = None,
+        g: float = 0.0,
+    ):
+
+        self.weighted = weighted
+        self.derivative = derivative
+        self.window = window
+        self.itakura_max_slope = itakura_max_slope
+        self.bounding_matrix = bounding_matrix
+        self.g = g
+
+        if not weighted and not derivative:
+            metric_key = "dtw"
+        elif not weighted and derivative:
+            metric_key = "ddtw"
+        elif weighted and not derivative:
+            metric_key = "wdtw"
+        elif weighted and derivative:
+            metric_key = "wddtw"
+
+        self.metric_key = metric_key
+
+        kwargs = {
+            "window": window,
+            "itakura_max_slope": itakura_max_slope,
+            "bounding_matrix": bounding_matrix,
+        }
+
+        # g is used only for weighted dtw
+        if weighted:
+            kwargs["g"] = g
+
+        self.kwargs = kwargs
+
+        super(DtwDist, self).__init__()
+
+    def _transform(self, X, X2=None):
+        """Compute distance/kernel matrix.
+
+            Core logic
+
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2
+                if X2 is not passed, is equal to X
+                if X/X2 is a pd.DataFrame and contains non-numeric columns,
+                    these are removed before computation
+
+        Parameters
+        ----------
+        X: 3D np.array of shape [num_instances, num_vars, num_time_points]
+        X2: 3D np.array of shape [num_instances, num_vars, num_time_points], optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X[i] and X2[j]
+        """
+        metric_key = self.metric_key
+        kwargs = self.kwargs
+
+        distmat = pairwise_distance(X, X2, metric=metric_key, **kwargs)
```

Comment:
it seems like a lot of boilerplate code to just call one line... any reason to go through the hassle?

Reply:
the one-line distances do not expose parameters in `sklearn` compatible style. If you want to use them in composition patterns like "k-nearest neighbors with that distance" and tune the distance parameters, you need to have the boilerplate wrap.

One *could* have implemented the distances right away like this, but the TSC crowd prefers the non-`sklearn` compatible interface. Which can be argued for, since it gives you sth that's more lightweight at a substantial cost of interoperability.

Reply:
e.g., the lightweight functions don't know which inputs are parameters that need be exposed via `sklearn`-style `get_params`, and which ones are data inputs, e.g., `X`, `X2`.

Reply:
I'm sure I'm missing something here since `pairwise_distance` if from sklearn so it is sklearn compatible. Is there an example workflow that you are trying to enable? That would help to understand where this is going.

Reply:
no, the `pairwise_distance` is from `sktime`, the `distances` module?
For the bigger picture of how to slot distances in estimators or aligners in distances and all to be composable with exposed parameters via `get_params`/`set_params` see the STEP here: https://github.com/sktime/enhancement-proposals/blob/main/steps/10_alignment_warping_knn/step.md

Reply:
OK, it seems I misread the imports. BTW sklearn has `pairwise_distances` not pariwise_distance` :sweat_smile: Thanks for the additional context!